app/javascript/packs/colorPicker.jsx

```diff
@@ -0,0 +1,28 @@
+import { h, render } from 'preact';
+import { ColorPicker } from '@crayons';
+
+// Find any color picker inputs on the page and replace them with the Preact enhanced component
+const colorInputs = document.querySelectorAll('[data-color-picker]');
+
+for (const input of colorInputs) {
+  const { labelText } = input.dataset;
+  const inputProps = {};
+
+  // Copy any specific attributes to the new input
+  const { attributes: inputAttributes } = input;
+  for (const attr of inputAttributes) {
+    inputProps[attr.name] = attr.value;
+  }
+
+  render(
+    <ColorPicker
+      id={inputProps.id}
+      defaultValue={input.value}
+      inputProps={inputProps}
+      buttonLabelText={labelText}
+    />,
+    input.parentElement,
+    input,
```

Comment:
Refactor (blocking): the third optional parameter is not necessary as Preact is unable to do a VDOM diff that updates the current input.

Reply:
@nickytonline Are you sure this needs to be blocking? That argument is [documented as `replaceNode`](https://preactjs.com/guide/v8/api-reference/#preactrender), and replacing the node appears to be what is intended here.

Reply:
It does replace the node, but only if it is able to do a virtual DOM diff where it can be updated otherwise the original node represent by the `replaceNode` optional parameter remains. In this case, the original element is an `<input />`, but the Preact component rendered is a `<div />` that contains an `<input />` and some other goodies.

![CleanShot 2022-02-16 at 17 23 50](https://user-images.githubusercontent.com/833231/154367827-1ab76885-2141-4de6-82f1-0044ccbd18df.png)

Here it is in action without the `input.remove()`

https://user-images.githubusercontent.com/833231/154366341-5c4d9727-960d-479a-b87a-835618933d8a.mp4

`input.remove()` is what removes it which is why I was requesting the `replaceNode` parameter to be removed.

From the creator of Preact in [this comment](https://github.com/preactjs/preact/issues/2004#issuecomment-573134053):

> Now, here's the case that doesn't work in Preact X, and that could potentially represent a bug that could manifest itself in other documented use-cases. If we want to do "pinpoint" rendering like this, but the new VDOM tree we pass to render() has a different root element type (eg: <a> vs <span>), Preact X will not remove the previous DOM element we passed and will instead prepend the newly created tree before it. This is because render()'s replaceNode parameter was designed to allow diffing Preact trees within a parent that has other elements Preact shouldn't touch.

Reply:
@nickytonline Fascinating, I hate it. :joy:

I mean, I like Preact, but I imagine the fact that this behavior isn't in the documentation probably leads to bugs and confusion. Also, I linked to the wrong version of the documentation above, but even [the current version](https://preactjs.com/guide/v10/api-reference/#render) doesn't make this behavior clear.